pkg/controller/apiextensions/composite/api.go

```diff
@@ -125,6 +151,93 @@ func (r *APISelectorResolver) ResolveSelector(ctx context.Context, cp resource.C
 	return errors.New(errNoCompatibleComposition)
 }
 
+// NewAPIDefaultCompositionSelector returns a APIDefaultCompositionSelector.
+func NewAPIDefaultCompositionSelector(c client.Client, ref v1.ObjectReference) *APIDefaultCompositionSelector {
+	return &APIDefaultCompositionSelector{client: c, defRef: ref}
+}
+
+// APIDefaultCompositionSelector selects the default composition referenced in
+// the definition of the resource if neither a reference nor selector is given
+// in composite resource.
+type APIDefaultCompositionSelector struct {
+	client client.Client
+	defRef v1.ObjectReference
+}
+
+// SelectComposition selects the default compositionif neither a reference nor
+// selector is given in composite resource.
+func (r *APIDefaultCompositionSelector) SelectComposition(ctx context.Context, cp resource.Composite) error {
+	if cp.GetCompositionReference() != nil || cp.GetCompositionSelector() != nil {
+		return nil
+	}
+	def := &v1alpha1.InfrastructureDefinition{}
+	if err := r.client.Get(ctx, meta.NamespacedNameOf(&r.defRef), def); err != nil {
+		return errors.Wrap(err, errGetInfraDef)
+	}
+	if def.Spec.DefaultCompositionRef == nil {
+		return nil
+	}
```

Comment:
Could we pass the default and enforced compositions down from the InfraDef controller to its child composite controllers, rather than having the child controllers query their parent InfraDef on every reconcile? Presumably this would require us to handle the case in which the default/enforced composition changed - i.e. by restarting the composition controllers.

This leads me to wonder whether we should support the enforced composition changing at all. Would changing the enforced composition [retcon](https://en.wikipedia.org/wiki/Retroactive_continuity) any existing composite resources to use the newly enforced composition?

Reply:
That was my initial approach, too; pass the `InfrastructureDefinition` object to both composition selectors instead of a reference to it. I see no problem making enforced composition ref `+immutable` but I believe we should support changing default composition ref. Not so sure how restarting the controller at every change would look like.

Right now, it's only the enforced selector that queries the infradef at every reconcile (default selector doesn't query anything after the first pass) because it cannot know whether the reference on the composite resource drifted or not without knowing what the enforced value is. The safe but less-performant approach is the current one. But I believe what we can do is to mark the enforced composition ref as `+immutable` and make `NewAPIEnforcedCompositionSelector` accept the enforced composition reference directly, relying on the fact that it never changes. It'd be a weak guarantee for now but when we get the `+immutable` webhook implemented, it'd be a strong guarantee.